### ISSUE TITLE
Updating all is breaking conda on Windows

### ISSUE THREAD

**Issue by gillins** [conda/conda#2846]
If I download a brand new `miniconda` for Python 3.5 64bit an do a `conda update --all`, `conda` appears broken:
```
c:\dev>conda
Traceback (most recent call last):
  File "C:\Users\sam\Miniconda3\Scripts\conda-script.py", line 5, in <module>
    sys.exit(main())
  File "C:\Users\sam\Miniconda3\lib\site-packages\conda\cli\main.py", line 94, i
n main
    imported = importlib.import_module(module)
  File "C:\Users\sam\Miniconda3\lib\importlib\__init__.py", line 126, in import_
module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 986, in _gcd_import
  File "<frozen importlib._bootstrap>", line 969, in _find_and_load
  File "<frozen importlib._bootstrap>", line 958, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 673, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 662, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "C:\Users\sam\Miniconda3\lib\site-packages\conda\cli\main_list.py", line
18, in <module>
    from ..egg_info import get_egg_info
  File "C:\Users\sam\Miniconda3\lib\site-packages\conda\egg_info.py", line 15, i
n <module>
    from .misc import rel_path
  File "C:\Users\sam\Miniconda3\lib\site-packages\conda\misc.py", line 18, in <m
odule>
    from .fetch import fetch_index
  File "C:\Users\sam\Miniconda3\lib\site-packages\conda\fetch.py", line 26, in <
module>
    from .connection import CondaSession, unparse_url, RETRIES
  File "C:\Users\sam\Miniconda3\lib\site-packages\conda\connection.py", line 51,
 in <module>
    python=platform.python_implementation(),
  File "C:\Users\sam\Miniconda3\lib\platform.py", line 1287, in python_implement
ation
    return _sys_version()[0]
  File "C:\Users\sam\Miniconda3\lib\platform.py", line 1250, in _sys_version
    repr(sys_version))
ValueError: failed to parse CPython sys.version: '3.5.1 |Continuum Analytics, In
c.| (default, Feb 16 2016, 09:49:46) [MSC v.1900 64 bit (AMD64)]'
```
Not sure if this is a problem with `conda-forge` Python, or `conda`. I have `conda-forge` as one of my channels.

**Comment by msarahan** [comment]
might be a conda-forge issue.  but the valueerror there confuses me greatly.  what's the output of ```conda list```?

**Comment by gillins** [comment]
Ummm I can't run that at the moment :smiley: Is there another way to find this information out?

**Comment by msarahan** [comment]
ugh... sorry.  zip up the conda-meta folder and post it somewhere?

**Comment by gillins** [comment]
Is a directory listing enough?
```
c:\dev>dir c:\Users\sam\Miniconda3\conda-meta
 Volume in drive C has no label.
 Volume Serial Number is 34FD-BA4C

 Directory of c:\Users\sam\Miniconda3\conda-meta

06/24/2016  09:59 AM    <DIR>          .
06/24/2016  09:59 AM    <DIR>          ..
06/24/2016  09:49 AM             6,823 conda-4.1.3-py35_0.json
06/24/2016  09:49 AM             7,124 conda-build-1.21.1-py35_0.json
06/24/2016  09:49 AM             4,260 conda-env-2.5.1-py35_0.json
06/24/2016  09:59 AM             1,566 history
06/24/2016  09:59 AM             3,968 jinja2-2.8-py35_0.json
06/24/2016  09:59 AM             2,259 markupsafe-0.18-py35_0.json
06/24/2016  09:56 AM             2,131 menuinst-1.4.1-py35_0.json
06/24/2016  09:49 AM               636 patch-2.5.9-1.json
06/24/2016  09:59 AM            36,994 pip-8.1.2-py35_0.json
06/24/2016  09:59 AM             3,838 psutil-4.1.0-py35_0.json
06/24/2016  09:59 AM             1,128 pycosat-0.6.1-py35_0.json
06/24/2016  09:59 AM            18,959 pycrypto-2.6.1-py35_0.json
06/24/2016  09:59 AM           218,235 python-3.5.1-0.json
06/24/2016  09:59 AM             2,862 pyyaml-3.11-py35_0.json
06/24/2016  09:59 AM            13,646 requests-2.9.2-py35_0.json
06/24/2016  09:49 AM             3,778 ruamel_yaml-0.11.7-py35_0.json
06/24/2016  09:59 AM               936 setuptools-23.0.0-py35_0.json
06/24/2016  09:59 AM             5,203 vs2015_runtime-14.0.25123-0.json
06/24/2016  09:48 AM             5,390 wheel-0.29.0-py35_0.json
06/24/2016  09:59 AM             1,026 yaml-0.1.6-vc14_0.json
              20 File(s)        340,762 bytes
               2 Dir(s)  28,668,203,008 bytes free
```

**Comment by msarahan** [comment]
that doesn't tell me where each package came from, unfortunately.  The history file there might be enough.

**Comment by gillins** [comment]
```
==> 2016-06-24 09:49:29 <==
# cmd: C:\Users\sam\Miniconda3\Scripts\conda-script.py install conda-build
conda-4.0.5-py35_0
conda-env-2.4.5-py35_0
menuinst-1.3.2-py35_0
pip-8.1.1-py35_1
pycosat-0.6.1-py35_0
pycrypto-2.6.1-py35_3
python-3.5.1-4
pyyaml-3.11-py35_3
requests-2.9.1-py35_0
setuptools-20.3-py35_0
vs2015_runtime-14.00.23026.0-0
wheel-0.29.0-py35_0
==> 2016-06-24 09:49:47 <==
# cmd: C:\Users\sam\Miniconda3\Scripts\conda-script.py install conda-build
-conda-4.0.5-py35_0
-conda-env-2.4.5-py35_0
+conda-4.1.3-py35_0
+conda-build-1.21.1-py35_0
+conda-env-2.5.1-py35_0
+jinja2-2.8-py35_1
+markupsafe-0.23-py35_2
+patch-2.5.9-1
+psutil-4.3.0-py35_0
+ruamel_yaml-0.11.7-py35_0
# install specs: ['conda-build', 'conda', 'python 3.5*']
==> 2016-06-24 09:59:13 <==
# cmd: C:\Users\sam\Miniconda3\Scripts\conda-script.py update --all
-jinja2-2.8-py35_1
-markupsafe-0.23-py35_2
-menuinst-1.3.2-py35_0
-pip-8.1.1-py35_1
-psutil-4.3.0-py35_0
-pycosat-0.6.1-py35_0
-pycrypto-2.6.1-py35_3
-python-3.5.1-4
-pyyaml-3.11-py35_3
-requests-2.9.1-py35_0
-setuptools-20.3-py35_0
-vs2015_runtime-14.00.23026.0-0
+conda-forge::jinja2-2.8-py35_0
+conda-forge::markupsafe-0.18-py35_0
+conda-forge::psutil-4.1.0-py35_0
+conda-forge::pycosat-0.6.1-py35_0
+conda-forge::pycrypto-2.6.1-py35_0
+conda-forge::python-3.5.1-0
+conda-forge::pyyaml-3.11-py35_0
+conda-forge::requests-2.9.2-py35_0
+conda-forge::vs2015_runtime-14.0.25123-0
+conda-forge::yaml-0.1.6-vc14_0
+menuinst-1.4.1-py35_0
+pip-8.1.2-py35_0
+setuptools-23.0.0-py35_0
```

**Comment by msarahan** [comment]
yeah, you got conda forge's python, but seeing as your error is still showing Continuum's tagline, something got botched in the replacement of files.  Sorry, this is more than I can chew right now.  @kalefranz - please investigate.

**Comment by gillins** [comment]
Thanks for looking into this. Let me know if you need any more info....

**Comment by mcg1969** [comment]
Googling "ValueError: failed to parse CPython sys.version" (with quotes) yields a fair amount of material. I'm putting my money on a problem with the conda-forge Python build at this point.

**Comment by mcg1969** [comment]
More here: http://stackoverflow.com/questions/34145861/valueerror-failed-to-parse-cpython-sys-version-after-using-conda-command

**Comment by mcg1969** [comment]
I was able to reproduce the error.

**Comment by mcg1969** [comment]
This is very easy to reproduce. Just install the latest released Miniconda for Python 3.5, then do

`conda install conda-build`
`conda update --all -c conda-forge`

What I believe is happening is that conda is successfully replacing the file `lib/platform.py` using the conda-forge version of the package, but it is failing to replace `python35.dll`. If you look at the file sizes for `python35.dll` in the Miniconda home directory and in `pkgs\python-3.5.1-0\`, you'll see they are different.

`lib/platform.py` contains a regex that parses `sys.version` to extract key elements. I'm not sure if Anaconda's Python modifies it, but conda-forge's does, and it doesn't match the `sys.version` string returned by Continuum's version of Python. Conda-forge's regex is
```
_sys_version_parser = re.compile(
    r'([\w.+]+)\s*\ \|\ packaged\ by\ conda\-forge\ \|\s*'
    '\(#?([^,]+),\s*([\w ]+),\s*([\w :]+)\)\s*'
    '\[([^\]]+)\]?', re.ASCII)
```
while Continuum's is
```
_sys_version_parser = re.compile(
    r'([\w.+]+)\s*'
    '(?:\|[^|]*\|)?\s*' # version extra
    '\(#?([^,]+),\s*([\w ]+),\s*([\w :]+)\)\s*'
    '\[([^\]]+)\]?', re.ASCII)
```
Bottom line: installing `conda-forge::python-3.5.1-0` on Windows, into the root environment at least, doesn't seem to actually *give* you that package, at least not fully.

@kalefranz @mingwandroid

**Comment by mcg1969** [comment]
Here is the `---debug` output of the last command:

https://gist.github.com/anonymous/48c82f00763097840d358f48d3d892b3

Search with your favorite text editor for `python35.dll` and enjoy. Or, weep bitterly. Lots of permissions-related warnings. Here is the nut graph:
```
DEBUG:conda.install:could not remove file: 'C:\Users\mcg_000\Miniconda3\python35.dll'
DEBUG:conda.install:Could not move C:\Users\mcg_000\Miniconda3\python35.dll to trash
WARNING:conda.install:file already exists: 'C:\\Users\\mcg_000\\Miniconda3\\python35.dll'
ERROR:conda.install:failed to unlink: 'C:\\Users\\mcg_000\\Miniconda3\\python35.dll'
DEBUG:conda.install:Could not move C:\Users\mcg_000\Miniconda3\python35.dll to trash
ERROR:conda.install:failed to link (src='C:\\Users\\mcg_000\\Miniconda3\\pkgs\\python-3.5.1-0\\python35.dll', dst='C:\\Users\\mcg_000\\Miniconda3\\python35.dll', type=1, error=OSError('win32 hard link failed',))
```
So it is confirmed: _conda is failing to move the old python DLL out of the way and failing to move the new one into place._

**Comment by mingwandroid** [comment]
Windows can't delete/rename/overwrite files that it are currently open. How this ever worked is beyond my understanding (I'd hazard that it works only sometimes). `move_to_trash` is a band-aid around the core problem that IMHO `conda` must be made into a standalone executable, perhaps using `py2exe` (called something like `conda-internal.exe`) with a special stub executable (called `conda.exe`) who's only purpose is to defer to `conda-internal.exe` and to manage updating `conda-internal.exe` when `conda-internal.exe` indicates this is necessary.

**Comment by mcg1969** [comment]
Clearly there is a major technical challenge to navigate here. But step 1, which in my view must be done immediately, is to prevent `conda` from _claiming_ it has installed a package that it has not. A fatal error is preferable here.

But @mingwandroid, can you please look at the _entire_ log? Do search for `python35.dll` but also look at everything around it. There are _lots_ of permissions issues here that do not seem to be explained by the fact that the files are already open---files that are already in the trash, for instance. Can we figure out what's going on there?

**Comment by mcg1969** [comment]
Also relevant: http://stackoverflow.com/a/7148939/1599393

**Comment by mcg1969** [comment]
@gillins, my recommendation here is not to attempt to update Python in the root environment until this can be sorted out.

**Comment by kalefranz** [comment]
It's too early in the day to have a headache like this (that's not from a hangover).

**Comment by mcg1969** [comment]
In my view, the first order of business is to understand why there are so many permissions errors in the log _that cannot be explained by the open files theory_.

**Comment by kalefranz** [comment]
Here's a real and possibly related error:
https://ci.appveyor.com/project/ContinuumAnalytics/conda/build/1.0.359/job/2jfyo9kobxs8je4p#L493

In this PR
https://github.com/conda/conda/pull/2853
I'm asking for the conda-forge pycosat to clobber the defaults pycosat in the package cache.  But it's refusing.

**Comment by mcg1969** [comment]
Time for an experiment:

- Fresh install of Miniconda (python 3.5)
- `conda --debug install -y -f python=3.5.1=2 msvc_runtime=1.0.1=vc14_0`
- `fsutil hardlink list Miniconda3\python.exe`
- `fsutil hardlink list Miniconda3\python35.dll`

The only reason I used `-f` here is to bypass conda's auto-update, so I could installl *just* the two packages needed downgrade python. You can try it without the `-f` if you want to confirm that this is the right recipe for the job. Lots of warnings about being unable to remove the Python executables and DLLs, and empty the trash (permission denied, then not empty). And yet, _the link step succeeds_. The `fsutil` commands confirm that I am indeed hard linking to the right version of Python.

Here's the output of that conda command: https://gist.github.com/51a1bd0babb4afa25711c198dd770908

**Comment by mcg1969** [comment]
Continuing my experiment:

- `conda install -f conda conda-env ruamel_yaml`
- `conda --debug update python`

So that `update` command is using the new version of `conda`. It *says* it is going to replace the conda-forge Python I just installed above with defaults.

It doesn't.
```
Python 3.5.1 | packaged by conda-forge | (default, Jun 15 2016, 12:45:45) [MSC v.1900 64 bit (AMD64)] on win32
```
and from the log:
```
ERROR:conda.install:failed to link (src='C:\\Users\\mcg_000\\Miniconda3\\pkgs\\python-3.5.1-4\\python35.dll', dst='C:\\Users\\mcg_000\\Miniconda3\\python35.dll', type=1, error=OSError('win32 hard link failed',))
```
So regardless of the validity of @mingwandroid's claim above---and believe me, I am inclined to believe him---*somehow*, conda 4.0.x is installing python over itself, and conda 4.1.x is not.

Here is the full conda log for this one: https://gist.github.com/c2ae0a40e46ed8fa19c73b732b8c9688

**Comment by mcg1969** [comment]
OK, this is intriguing. The above experiments suggest that conda 4.0.x can install python on top of itself, while conda 4.1.x cannot. Well, it's not that simple: I just did this:

- `conda install conda=4.0.9`
- `conda --debug update -f python=3.5.1=2 -y`

And sure enough, conda 4.0.9 _gives the same "failed to link" error as above._ 
https://gist.github.com/5adbe66eb39c809a4bb1020d5ffa1332

There is some sort of underlying permissions problem happening here, _caused_ by something in 4.1, but _persists_ even when you downgrade.

**Comment by kalefranz** [comment]
I smell the fishiness too. Here and in other places. Something's just not quite right. 🤔

**Comment by mcg1969** [comment]
Downgrading to 3.19.0 enabled me to install `python=3.5.1=4`

**Comment by kalefranz** [comment]
CC @csoja

**Comment by mcg1969** [comment]
Here's the update. Bisection revealed this commit as the culprit: https://github.com/conda/conda/commit/226ac99f07e8394e025de6f18be86f2a927d3c1c

The PR is #2542, the issue is https://github.com/conda/conda-build/issues/891.

I've confirmed that functionally reverting this commit fixes the problem for 4.1.3. However:
- _It makes no sense_. This PR concerns itself only with the `move_to_trash` functionality. The failure here is in the _link_ step, which occurs afterwards. There is an important but subtle difference, apparently, between UNC paths and Windows paths, that has persistent side effects.
- This means issue #891 will break again. I see no choice but to revert this commit, but we need an alternative.
